Use glClear to clear the stencil buffer.

colesbury opened the pull request:
Use glClear instead of drawing a rectange so that current layer's mask
for the entire stencil buffer is cleared. This fixes clipping when
drawn to a texture larger than the screen size or when the stencil or
clipping node's children contain vertices with negative coordinates.

Fixes issue 1463  http://code.google.com/p/cocos2d-iphone/issues/detail?id=1463
(CCClippingNode does not clear the stencil buffer properly)

I've checked the clipping tests and they look ok.

colesbury committed:
Use glClear to clear the stencil buffer.

Use glClear instead of drawing a rectange so that current layer's mask
for the entire stencil buffer is cleared. This fixes clipping when
drawn to a texture larger than the screen size or when the stencil or
clipping node's children contain vertices with negative coordinates.

pierredavidbelanger commented:
I understand that http://code.google.com/p/cocos2d-iphone/issues/detail?id=1463 need to be fix, but I think it will not be a good idea to revert using glClear here.
This is what I was doing in the first implementation of CCClippingNode, and it did not work on some recent MBP hardware.

Here is the commit where I swapped the glClear method for a fullscreen rectangle:
https://github.com/cocos2d/cocos2d-iphone/commit/1611da9702d030f5a9767fd7a02ce8ace959b0bd#cocos2d/CCClippingNode.m

colesbury commented:
Thanks for your quick reply!  I'd love to get this fixed.

Do you remember which tests didn't work on MBP when using glClear?

pierredavidbelanger commented:
Here is what the "Raw Stencil Tests" looks like on OS X (OS X 10.7.4, AMD Radeon HD 6750M) with the glClear:

https://github.com/cocos2d/cocos2d-iphone/pull/252#issuecomment-9821861

I, my self, was not able to reproduce this problem. @ricardoquesada helped me debug and fix the issue.

colesbury committed:
Use ccDrawRect to clear stencil buffer on OS X.

Works around a bug in some ATI drivers on 10.7 where glStencilMask
does not affect glClear.

Add a test where the stencil is positioned below the clipping node's origin.

colesbury commented:
Sounds like it's important to work around that AMD driver bug.  I've updated the pull request to use ccDrawRect on OS X and added a test case that demonstrates the original problem I was having.

ricardoquesada commented:
Thanks @colesbury 

@pierredavidbelanger If you agree with the patch, I will merge it.

pierredavidbelanger commented:
I tested the patch in the simulator, an iPhone 4, an iPad 2 and a mid-2009 MBP, everything looks good.
Except that I can not confirm that the bug with glClear on some ATI drivers is still inexistant; but by reading the code, we can tell it should still be ok.

I think it is a good idea to revert using glClear on iOS.

Good job @colesbury

pierredavidbelanger commented:
Oups, obviously that the bug with glClear on OSX is still inexistant: glClear is called only when on iOS ;)

ricardoquesada commented:
Thanks .I will merge it.
